libsolidity/formal/BMC.cpp

```diff
@@ -290,104 +291,214 @@ bool BMC::visit(Conditional const& _op)
 	return false;
 }
 
-// Here we consider the execution of two branches:
-// Branch 1 assumes the loop condition to be true and executes the loop once,
-// after resetting touched variables.
-// Branch 2 assumes the loop condition to be false and skips the loop after
-// visiting the condition (it might contain side-effects, they need to be considered)
-// and does not erase knowledge.
-// If the loop is a do-while, condition side-effects are lost since the body,
-// executed once before the condition, might reassign variables.
-// Variables touched by the loop are merged with Branch 2.
+// Unrolls while or do-while loop
 bool BMC::visit(WhileStatement const& _node)
 {
-	auto indicesBeforeLoop = copyVariableIndices();
+	auto indicesBefore = copyVariableIndices();
+
+	m_context.pushSolver();
+	// variables touched by loop might change their value
+	// assume their values are not constant
 	m_context.resetVariables(touchedVariables(_node));
-	decltype(indicesBeforeLoop) indicesAfterLoop;
+
+	_node.condition().accept(*this);
+
+	// do not apply condition target on expressions
+	// that value might be changed by other loops
+	if (isRootFunction() && (!isInsideLoop()))
+		addVerificationTarget(
+			VerificationTargetType::ConstantCondition,
+			expr(_node.condition()),
+			&_node.condition()
+		);
+	m_context.popSolver();
+
+	resetVariableIndices(indicesBefore);
+
+	unsigned int start = 0;
+	smtutil::Expression broke(false);
 	if (_node.isDoWhile())
 	{
-		indicesAfterLoop = visitBranch(&_node.body()).first;
-		// TODO the assertions generated in the body should still be active in the condition
-		_node.condition().accept(*this);
-		if (isRootFunction())
-			addVerificationTarget(
-				VerificationTargetType::ConstantCondition,
-				expr(_node.condition()),
-				&_node.condition()
-			);
+		loopScopes.emplace();
+
+		auto indicesAfter = visitBranch(&_node.body()).first;
+
+		smtutil::Expression continues(false);
+		for (auto const& loopControl: loopScopes.top())
+		{
+			if (loopControl.kind == LoopControlKind::Break)
+			{
+				mergeVariables(
+					!broke && !continues && loopControl.pathConditions,
+					loopControl.variableIndicies,
+					copyVariableIndices()
+				);
+				broke = broke || loopControl.pathConditions;
+			}
+			else if (loopControl.kind == LoopControlKind::Continue)
+			{
+				mergeVariables(
+					!broke && !continues && loopControl.pathConditions,
+					loopControl.variableIndicies,
+					copyVariableIndices()
+				);
```

Comment:
Similarly here, the call to `mergeVariables` can be extracted before the `if/else`.